src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPoolManager.cs

```diff
@@ -72,17 +68,85 @@ public HttpConnectionPools(HttpConnectionSettings settings, bool usingProxy)
 
-        /// <summary>Gets a pool for the specified endpoint, adding one if none existed.</summary>
-        /// <param name="key">The endpoint for the pool.</param>
-        /// <returns>The retrieved pool.</returns>
-        public HttpConnectionPool GetOrAddPool(HttpConnectionKey key)
+        private static string ParseHostNameFromHeader(string hostHeader)
+        {
+            // See if we need to trim off a port.
+            int colonPos = hostHeader.IndexOf(':');
+            if (colonPos >= 0)
+            {
+                // There is colon, which could either be a port separator or a separator in
+                // an IPv6 address.  See if this is an IPv6 address; if it's not, use everything
+                // before the colon as the host name, and if it is, use everything before the last
+                // colon iff the last colon is after the end of the IPv6 address (otherwise it's a
+                // part of the address).
+                int ipV6AddressEnd = hostHeader.IndexOf(']');
+                if (ipV6AddressEnd == -1)
+                {
+                    return hostHeader.Substring(0, colonPos);
+                }
+                else
+                {
+                    colonPos = hostHeader.LastIndexOf(':');
+                    if (colonPos > ipV6AddressEnd)
+                    {
+                        return hostHeader.Substring(0, colonPos);
+                    }
+                }
+            }
+
+            return hostHeader;
+        }
+
+        private static HttpConnectionKey GetConnectionKey(HttpRequestMessage request, Uri proxyUri)
         {
+            Uri uri = request.RequestUri;
+
+            string sslHostName = null;
+            if (HttpUtilities.IsSupportedSecureScheme(uri.Scheme))
+            {
+                string hostHeader = request.Headers.Host;
+                if (hostHeader == null)
+                {
+                    // No explicit Host header.  Use host from uri.
+                    sslHostName = uri.IdnHost;
+                }
+                else
+                {
+                    sslHostName = ParseHostNameFromHeader(hostHeader);
+                }
+            }
+
+            if (proxyUri != null)
+            {
+                Debug.Assert(HttpUtilities.IsSupportedNonSecureScheme(proxyUri.Scheme));
+                if (sslHostName != null)
+                {
+                    // Standard HTTP proxy usage for non-secure requests
+                    // The destination host and port are ignored here, since these connections
+                    // will be shared across any requests that use the proxy.
+                    return new HttpConnectionKey(null, 0, null, proxyUri);
+                }
+                else
+                {
+                    // Tunnel SSL connection through proxy to the destination.
+                    return new HttpConnectionKey(uri.IdnHost, uri.Port, sslHostName, proxyUri);
+                }
```

Comment:
Is this right?  Seems like the branches are swapped.

Reply:
Huh. It does seem wrong.  

Does that mean we don't have any test coverage of this case?

Reply:
To be clear, we *do* have test coverage.  The incorrect logic above screws up the key definition, but proxy connections are still handled properly.

Reply:
I added some asserts in HttpConnectionPool to check this.

Reply:
Thanks.